src/client/index.js

```diff
@@ -445,77 +448,143 @@ function updateLint(lint) {
     const simEngine = $('#engine').val();
-    destroycircuit();
-    $.ajax({
-        type: 'POST',
-        url: '/api/yosys2digitaljs',
-        contentType: "application/json",
-        data: JSON.stringify({ files: data, options: opts }),
-        dataType: 'json',
-        success: (responseData, status, xhr) => {
-            let circuit = responseData.output;
-            if (transform) circuit = digitaljs.transform.transformCircuit(circuit);
-            const engines = { synch: digitaljs.engines.BrowserSynchEngine, worker: digitaljs.engines.WorkerEngine };
-            mkcircuit(circuit, {layoutEngine: layoutEngine, engine: engines[simEngine]});
-            updateLint(responseData.lint);
-        },
-        error: (request, status, error) => {
-            loading = false;
-            updatebuttons();
-            $('form').find('input, textarea, button, select').prop('disabled', false);
+    const engines = { synch: digitaljs.engines.BrowserSynchEngine, worker: digitaljs.engines.WorkerEngine };
+
+    mkcircuit(transform ? digitaljs.transform.transformCircuit(circuit) : circuit, {layoutEngine: layoutEngine, engine: engines[simEngine]});
+    updateLint(lint);
+    openTab(circuitTabClass);
+}
+
+const synthesize = (function() {
+    let yosysWorker = null;
+
+    return function(onSuccess, onError) {
+        const data = {};
+        for (const [name, editor] of Object.entries(editors)) {
+            const panel = $('#' + name);
+            data[panel.data("filename") + "." + panel.data("extension")] = editor.getValue();
+            editor._is_dirty = false;
+        }
+        for (const [filename, file] of Object.entries(filedata)) {
+            data[filename] = file.result;
+        }
+        if (Object.keys(data).length == 0) {
             $('<div class="query-alert alert alert-danger alert-dismissible fade show" role="alert"></div>')
                 .append('<button type="button" class="close" data-dismiss="alert" aria-label="Close"><span aria-hidden="true">&times;</span></button>')
-                .append(document.createTextNode(request.responseJSON.error))
-                .append($("<pre>").text(request.responseJSON.yosys_stderr.trim()))
+                .append(document.createTextNode("No source files for synthesis."))
                 .appendTo($('#toolbar'))
                 .alert();
+            return;
+        }
+        const opts = {
+            optimize: $('#opt').prop('checked'),
+            fsm: $('#fsm').val(),
+            fsmexpand: $('#fsmexpand').prop('checked'),
+            lint: $('#lint').prop('checked')
+        };
+        destroycircuit();
+
+        const synthesisMode = $('#synthesis-mode').val();
+
+        const synthesisStrategies = {
+            wasm: (data, opts) => {
+                yosysWorker = yosysWorker ?? new Worker(new URL("worker.js", import.meta.url), {
+                    type: 'module'
+                });
+
+                yosysWorker.onmessage = function(e) {
```

Comment:
Why is `onmessage` redefined each time, even though `yosysWorker` is cached? It's because of `onSuccess` and `onError` parameters, right? There should be never more than one request in progress, so this is fine, but this looks weird and the assumption is not stated anywhere.